### ISSUE TITLE
Spatial file manager

### ISSUE THREAD

**Issue by probonopd** [helloSystem/hello#79]
The current https://github.com/helloSystem/Filer/ implementation is not spatial.

**Spatial Interfaces** are explained by John Siracusa at
https://arstechnica.com/gadgets/2003/04/finder/2/

It would be nice if we could have a spacial file manager.

**Comment by probonopd** [comment]
Also see https://www.asktog.com/columns/034OSX-FirstLook.html

**Comment by probonopd** [comment]
Possibly store the metadata for this using [extended attributes](https://www.freebsd.org/cgi/man.cgi?setextattr)?

**Comment by moochris** [comment]
Were you thinking that Filer should be extended to have a spatial mode or having an alternative to Filer that is spatial? I don't think there's any existing project in Qt (Gnome 2 had spatial mode at one time - we don't want that)

**Comment by probonopd** [comment]
Either way. One thing I don't like about Filer is that it still draws in some glib based components, and I was considering alternatives including DFilemanager but I could not get it to compile yet: https://github.com/helloSystem/Filer/issues/23

**Comment by moochris** [comment]
I have started extending Filer for spatial mode support - https://github.com/helloSystem/Filer/pull/35

I agree it would be nice not to have the glib components at all, but I think the current Filer is probably the best option still and for pragmatism it might be worth sticking with it and hope that eventually it can be refactored (well, libfm-qt)? I got DFilemanager to compile but there are more problems there.

**Comment by moochris** [comment]
Extended attributes looks viable - this is the next step in extending Filer. As a proof-of-concept from the command line:

```
HelloChris% setextattr user window.height 100 Documents
HelloChris% getextattr user window.height Documents 
Documents       100
```

See https://www.daemon-systems.org/man/setextattr.1.html

**Comment by moochris** [comment]
I had some spare time (I'm on 'holiday' this week :) ) so I went ahead with the extended attribute implementation for window position and size (modified the same PR).

This implements spatial folder browsing, so obviously there's an outstanding task for storing icon position as well but I think this is good to start with?

**Comment by probonopd** [comment]
Wow @moochris this is awesome! Thank you very much.

Let's think about a few more aspects of spatial navigation Maybe we should think about creating a spec for the metainformation needed to make a truly spatial file manager first, ideally in a way so that any file manager could implement the spec in a compatible way, e.g.,

* Window size in pixels at normal (no-HiDPI) resolution
* Window position on screen in a coordinate system
* Display mode (icons view, list view, Miller columns,...) and a definition of fallbacks (e.g., if Miller columns were set but the file manager does not have them, then use list view)
* Icon size for the window
* How icons are positioned (automatic/by type/sorted/manual...) and sort order
* Icon positions in a coordinate system for each icon
* Window background wallpaper (only in icon view)
* Tags/colors of files
* ...
As for storing the information, we should think about consider
* Extended Attributes (advantage: no additional files cluttered around for storing this information)
* One hidden file per directory (advantage: survives archiving and copy operations even over the network, etc.)
* Possibly a combination of the two or something else.

We should investigate how other spatial file managers are doing this.

**Comment by probonopd** [comment]
More food for thought from reviewer John Siracusa and Classic Mac UX auhtority Bruce "Tog" Tognazzini:

* https://arstechnica.com/gadgets/2003/04/finder/2/
* https://www.asktog.com/columns/034OSX-FirstLook.html

For me, the possibility to open new windows for each folder is just one (and probably the most controversial - hence Apple made it optional in Mac OS X) aspect. Other aspects should also be considered, such as remembering for each location in which mode (list/icons,...) it is displayed, where each icon is (in icon view mode), how things are sorted (in list view mode).

**Comment by moochris** [comment]
Yep I think a spec would be a good idea.

I started on a kind of structure for the names of the extended attributes in the initial implementation, but it probably won't scale to include the other fields.

I had thought of a few similar to what you have listed, but also another nice thing might be to reuse Haiku's HVIF format to store icons in extended attributes? (I'm sure you're aware of this, but https://www.haiku-os.org/articles/2006-11-13_why_haiku_vector_icons_are_so_small/ link anyway :smile: )

And yes storage is definitely a consideration - e.g. it would be nice if Nextcloud (or Dropbox or whatever) could synchronise the metadata. I quite like the extended attributes (seem to work well in this first attempt) but I guess if they are transparent to the user, then some other file based mechanism like dot files would be preferable.

**Comment by probonopd** [comment]
> I quite like the extended attributes (seem to work well in this first attempt) but I guess if they are transparent to the user, then some other file based mechanism like dot files would be preferable.

Let's have a look at where other systems are storing which kind of information; Haiku and the Mac are coming to mind immediately.

**Comment by moochris** [comment]
An interesting post (albeit from a while ago now) on extended attributes:

https://www.lesbonscomptes.com/pages/extattrs.html

Hopefully some of those pitfalls have been solved in the meantime. I note that it mentions ROX using them, so worth adding that to the list of projects to investigate.

**Comment by probonopd** [comment]
Possibly we need to consider a combination of extended attributes (for performance) and files (for robustness) (I _think_ this is what Haiku is doing). In any case, the spatial information should travel _with_ the folder, so that e.g., if someone accesses the same folder over the network the spatial information should be available to _that_ user, too.

**Comment by probonopd** [comment]
Need to keep in mind use cases like this one:

![](https://miro.medium.com/max/519/0*aU-yaB4t5HEwrPnc.png)

Here, whoever created that disk image
* Set the sidebar to disabled for that window
* Set a custom window wallpaper
* Set the size and location of the icons

...and all of it traveled along inside the disk image.

**Comment by moochris** [comment]
> Possibly we need to consider a combination of extended attributes (for performance) and files (for robustness) (I _think_ this is what Haiku is doing). In any case, the spatial information should travel _with_ the folder, so that e.g., if someone accesses the same folder over the network the spatial information should be available to _that_ user, too.

I think Haiku only uses attributes for Tracker - it's these here:
https://git.haiku-os.org/haiku/tree/src/kits/tracker/Attributes.h

e.g. `_trk/windframe`

But yes I agree these things should be portable. For Haiku, for example, they modified `zip` and `unzip` to preserve attributes. We should look into whether other formats can preseve extended attributes already - e.g. `tar` already does.

**Comment by probonopd** [comment]
Uh. I am not a huge fan of patching half the system. Maybe better use dot files in each directory. E.g, `.extattr`

**Comment by moochris** [comment]
Yeah, OK, I think you're right.

We'll need a spec - doesn't have to be fully fleshed out, but if we can get a standard for naming to start with that would be good for modifying the spatial Filer PR (as a prototype).

All files/folders will need a dot file I think, rather than a just one for the directory or you have all sorts of management of what then becomes like a flat database.

e.g.:

```
/home/chris/projects
/home/chris/.extattr_projects
/home/chris/projects/document.odt
/home/chris/projects/.extattr_document.odt
```
The dot files can contain key value pairs for the attributes, although I'm not sure about the best format - possibly just key=value one per line?

It doesn't need to apply the extended attributes at all really, as the dot files will need to be kept up to date - may as well _just_ use the dot files?

**Comment by moochris** [comment]
OK, I see the problem with that - the directory needs to contain its own dotfile or you'll lose it on archiving.

So:
```
/home/chris/projects
/home/chris/projects/.extattr
/home/chris/projects/document.odt
/home/chris/projects/.extattr_document.odt
```

**Comment by probonopd** [comment]
Instead of having

```
/home/chris/projects/.extattr_document.odt
```

we could have that information in 

```
/home/chris/projects/.extattr
```

**Comment by moochris** [comment]
It won't scale for very large numbers of files in a directory I think

**Comment by probonopd** [comment]
People are gonna hate this if we have such files cluttered around in big numbers.

Actually, let's have a look at how `.DS_Store` files work.

**Comment by probonopd** [comment]
- https://0day.work/parsing-the-ds_store-file-format/
- https://wiki.mozilla.org/DS_Store_File_Format
- http://search.cpan.org/~wiml/Mac-Finder-DSStore/DSStoreFormat.pod
- https://digi.ninja/projects/fdb.php

**Comment by moochris** [comment]
Looks a bit complicated!

**Comment by probonopd** [comment]
Our motto: Aim for making everything 10x as simple! But it cannot hurt to understand the concepts first.

**Comment by moochris** [comment]
> Sure. But we need to think about what to put where, so that it can travel along, e.g., in zip files and such. And we need to think about performance. The Wikipedia article has some good hints.

I'm suggesting that the single `.extattr` file in the directory is an sqlite database, so it would persist with the zipped directory. But yes, not sure about performance.

**Comment by probonopd** [comment]
Yes, that would be worth a try.

**Comment by probonopd** [comment]
Check this out!

https://ds-store.readthedocs.io/en/latest/

Manipulate Finder `.DS_Store` files from Python.

Wouldn't it rock if we could use the same format even, given that someone seems to already have written the implementation?

**Comment by antranigv** [comment]
Please let’s not use DS_Store, as a former macOS user it was a PITA for me, specially when doing development. 

This means it will be very useful for the daily users, but very bad experience for the power users and developers.

**Comment by probonopd** [comment]
The format seems to be documented here:
https://metacpan.org/pod/distribution/Mac-Finder-DSStore/DSStoreFormat.pod

What were your gribes with `.DS_Store` fies? I am asking because I fear that some of them will be there with `any` such format.

At the very least, we should _understand_ `.DS_Store` before we implement something new.

**Comment by probonopd** [comment]
At the bare minimum, https://metacpan.org/pod/distribution/Mac-Finder-DSStore/DSStoreFormat.pod tells us conceptually which type of information we should consider to store.

**Comment by antranigv** [comment]
My issue was not with DS_Store specifically, but with the concept that there’s a dot file in each directory which is tightly integrated with Finder.

If I did mv/cp manually from the CLI, then DS Store would not catch that. 
And I had to ignore it every time I did development either with git or svn. 

Again, I think for the end user it’s the best thing out there, but one level up (power-user) and it starts to be a pain.

**Comment by moochris** [comment]
So would extended attributes be better, even if you might lose spatial information on transfer to some file systems or network?

**Comment by moochris** [comment]
Note, that unlike with Haiku (who use their own BeFS attributes), we don't have to patch archivers and network file systems for support.

With extended attributes, it looks like tar, zip, NFS all have support - we are in a better situation than Haiku with this.

**Comment by antranigv** [comment]
> So would extended attributes be better, even if you might lose spatial information on transfer to some file systems or network?

I think that's fine, since that's how it work in other systems as well.

I have not done a full research here, but extended attributes seem very nice compared to having a dot file in every directory.

On the other hand, as @probonopd noted, we need to handle "image" installers as well, but that seems to be doable if we zip things, right? or we would loose the attributes if we distribute things over the net? hmm this needs trial and error to understand better.

**Comment by probonopd** [comment]
> My issue was not with DS_Store specifically, but with the concept that there’s a dot file in each directory

That will probably be unavoidable if we want metadata to travel in a **safe way** with a directory...

I know that the feature will not be for everyone so maybe we will make it possible to deactivate the file generation.

**Comment by moochris** [comment]
OK, so I think we should definitely have a dot file mechanism in addition to the extended attributes.

I tried the FreeBSD zip/unzip and it either doesn't have the patches or support for preserving the extended attributes.

Have a preference to turn on/off writing to the dot file (but always attempt to read/write to extended attributes). I will look at how the performance of sqlite is for this (next weekend maybe).

I think .DS_store files we could look at adding maybe just read support later on by reusing an existing parser somehow? Essentially, as legacy support.

Does that sound like a good plan @probonopd ?

**Comment by probonopd** [comment]
Indeed. Since poeple do not universally like the presence of `.DS_store` files I was thinking, why add yet another such file if we could use the ones that may already be there...

**Comment by probonopd** [comment]
We also need to do a systematic check whether all functionality can be reached in spatial mode, e.g., unmounting devices.

**Comment by bitigchi** [comment]
> We also need to do a systematic check whether all functionality can be reached in spatial mode, e.g., unmounting devices.

All mounted drives, devices, networks should be available through a 'Go' menu.

Also, please don't t do `.DS_Store` files. They are a PITA while dealing with `git` and other operating systems. I think Haiku uses extended attributes for the window properties, but if it's not possible, there are [talks](https://dev.haiku-os.org/ticket/10094) for a special workaround for the read-only folders. Maybe something similar can be utilised.

**Comment by moochris** [comment]
> We also need to do a systematic check whether all functionality can be reached in spatial mode, e.g., 
> * Unmounting devices
> * Going "up" in the hierarchy (is currently possible via the menu but we need Cmd-ArrowUp to work, and it should open the correct spatial window rather than changing the contents of the current one)

Yep, Haiku has an 'open parent' menu item which is also Cmd-ArrowUp.

It would be nice if all devices could show up on the desktop and be ejected from there (drag to trash?)

**Comment by moochris** [comment]
> Also, please don't do `.DS_Store` files. They are a PITA while dealing with `git` and other operating systems. I think Haiku uses extended attributes for the window properties, but if it's not possible, there are [talks](https://dev.haiku-os.org/ticket/10094) for a special workaround for the read-only folders. Maybe something similar can be utilised.

We are using extended attributes already (just window position and size at the moment). The dot file suggestion was because there are locations the attributes won't persist - e.g. some archives, network shares, cloud sync etc. And the plan is to use both but have an option to just write the attributes in the preferences, so you could turn off the annoying dot files.

I'm not 💯 convinced it's the right solution yet anyway. I was thinking that we could at least make whatever GUI archiving tool we supply preserve extattrs, and add requests that NextCloud handles them etc. 

The read only areas we already have an idea about how to solve, similar to the one suggested in that bug report.

**Comment by moochris** [comment]
@probonopd BTW, I raised another PR at the weekend for Linux support of the extended attributes:

https://github.com/helloSystem/Filer/pull/37

**Comment by probonopd** [comment]
> It would be nice if all devices could show up on the desktop and be ejected from there (drag to trash?)

Definitely. Check
https://github.com/helloSystem/Filer/issues/9
for some details.

Ejecting should also be possible via the right-click context menu.

**Comment by ctrlcctrlv** [comment]
I'm very excited for this. Back to basics, it'll be usable on Linux, right?

**Comment by moochris** [comment]
> I'm very excited for this. Back to basics, it'll be usable on Linux, right?

Yeah the spatial mode works on Linux too, which is nice because I use xubuntu on my dev machine for my day job. Obviously it's better in helloSystem! 😀

I'm not sure what @probonopd thinks about how integrated it should be with FreeBSD in future and as to whether certain features won't work on Linux.

**Comment by moochris** [comment]
I have created a new PR to fix the operation of all the items in the 'Go' and 'Bookmarks' menu now in spatial mode, so, e.g. alt-up will open parent in its own window (with remembered position/size):

https://github.com/helloSystem/Filer/pull/40

**Comment by probonopd** [comment]
Thank you very much @moochris, this is really great work. :+1:

**Comment by probonopd** [comment]
> e.g. alt-up will open parent in its own window 

Do you see Alt-Up there?

![image](https://user-images.githubusercontent.com/2480569/109430814-7157c480-7a03-11eb-9910-7e13f009cc32.png)

**Comment by moochris** [comment]
> > e.g. alt-up will open parent in its own window
> 
> Do you have a working shortcut there?

Yes - that's weird. Perhaps the translations have some bearing on the keyboard shortcuts? It's been a while since I looked at the Qt translation stuff.

![filerscreenshot](https://user-images.githubusercontent.com/11302084/109431124-d2b36f80-79d2-11eb-91c2-be413b47b11e.png)

**Comment by probonopd** [comment]
Yes, definitely; I need to copy the translations in place :)

**Comment by probonopd** [comment]
Currently those shortcuts do not appear/work when the language is set to German (`LANG=de_DE.UTF-8`); if I kill Filer and restart it with 

```
% LANG=C /System/Filer.AppDir/AppRun --desktop
```

then I *do* see those shortcuts (and they are working as intended).

All the shortcuts in Filer need to be reworked (and probably re-translated) sooner or later (e.g., Command-I for "Get Info" etc.), so probably we should not invest the time for looking into this before we have resolved https://github.com/helloSystem/hello/issues/46.

**Comment by moochris** [comment]
@probonopd this PR is my proposed solution to storing spatial metadata for the 'read only' parts of the filesystem, i.e. for areas with no user write permissions:

https://github.com/helloSystem/Filer/pull/46

If the extended attributes can't be written to a directory, then the directory with full path is created under ~/.config/hello/filer-metadata and the extattrs written there

Likewise, if an extattr is missing, the mirrored directory tree is checked as well.

The effect is that the window position and size can now be preserved and restored for e.g. the 'computer://' window and applications directory etc.

Obviously these attributes are only local to the computer you are using, but I think it's unlikely these read only areas will be copied around?

**Comment by probonopd** [comment]
hello @moochris.

Storing a file in ~ can allow one to store spatial information for read-only directories, but twhat happens if
* The information in the extended attributes of the directory and the information in the file dffer
* The directory gets deleted

Also, let's use a file in `~/.config/filer-qt/default/...`.

Can the same file format be used to store a file inside a directory, so that metadata can be preserved in situations in which entended attributes aren't?

I would see extended attributes more like a "cache" of sorts, with the information in the files being authoritative. Does that make sense?

**Comment by moochris** [comment]
> Storing a file in ~ can allow one to store spatial information for read-only directories, but twhat happens if
> * The information in the extended attributes of the directory and the information in the file dffer

The directory in the file system is always read first and the one under ~ is only read if the extattr doesn't exist.

> * The directory gets deleted

Then we are left with an empty directory under ~. Not a nice thought in terms of 'tidyness' but not really a space concern? We could always start periodically clearing out any that are orphaned.

> Also, let's use a file in `~/.config/filer-qt/default/...`.

Ok, will amend

> Can the same file format be used to store a file inside a directory, so that metadata can be preserved in situations in which entended attributes aren't?

It's not a file storing it in some other format, it's a tree of directories with the extattrs applied... At the moment we only have folder spacial information stored; when we extend to files then we can use empty files with extattrs to do the same thing?

> I would see extended attributes more like a "cache" of sorts, with the information in the files being authoritative. Does that make sense?

Yeah but all we have so far is extended attributes - I haven't added any dot files or anything yet

**Comment by moochris** [comment]
So for reading the extattrs, perhaps a better logic is:
if the directory is writable for the user, then read the extattrs directly; otherwise, read from the mirror under ~

The writing logic is sound as it is I think (it writes to the mirror under ~ if it can't write to the location specified)

**Comment by probonopd** [comment]
Ah, you are using empty directories to store the information.
Since we are going to need some form of dot files sooner or later anyway (if we decide against parsing the `.DS_Store` files) I was thinking that we could use the same file format for storing the information within ~. Wdyt?

**Comment by moochris** [comment]
Yes, I think we do the same for the dot files and store within the directories under `~` for the non-user-writable paths

**Comment by probonopd** [comment]
Found a Golang implementation but not a C/C++ one yet. I wonder whether we should just call a binary to read/write `.DS_Store:
https://github.com/gwend/dsstore

**Comment by moochris** [comment]
That's interesting, certainly worth testing with.

I guess the reason I'm a bit unsure about using them as the 'main' dot file mechanism is whether these reverse engineered parsers will be fast enough - it's one thing to write a tool that can read/write these as a command to play about with, but another to have an implementation usable for actually dealing with metadata on the fly in a file manager.

**Comment by probonopd** [comment]
We read this only when the folder is opened - how many folders does a user usually open at once? I'd say anything under [50 milliseconds](https://asktog.com/atc/principles-of-interaction-design/) would probably be acceptable.

**Comment by probonopd** [comment]
Some performance tests with the Python version:

```
% sudo pip-3.7 install ds-store

% wget "https://github.com/gwend/dsstore/blob/master/testdata/00.DS_Store?raw=true" -O DS_Store

% time ds_store DS_Store
DS_Store

DS_Store

.                bwsp {
    'ContainerShowSidebar': False,
    'ShowPathbar': False,
    'ShowSidebar': False,
    'ShowStatusBar': False,
    'ShowTabView': False,
    'ShowToolbar': False,
    'SidebarWidthTenElevenOrLater': 284.0,
    'WindowBounds': '{{200, 458}, {360, 222}}'
}
.                icvp {
    'arrangeBy': 'none',
    'backgroundColorBlue': 1.0,
    'backgroundColorGreen': 1.0,
    'backgroundColorRed': 1.0,
    'backgroundImageAlias': b'\x00\x00\x00\x00\x01\xba\x00\x02\x00\x00\x0cMacin'
                            b'tosh HD\x00\x00\x00\x00\x00\x00\x00\x00\x00'
                            b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00BD'
                            b'\x00\x01\xff\xff\xff\xff\x14Background_Black.'
                            b'png\x00\x00\x00\x00\x00\x00\x00\x00\x00'
                            b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
                            b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
                            b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xff\xff'
                            b'\xff\xff\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
                            b'\x00\x00\xff\xff\xff\xff\x00\x00\n cu'
                            b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
                            b'\x00\tGetscreen\x00\x00\x02\x00Y/:Users:gwend:Li'
                            b'brary:Mobile Documents:com~apple~CloudDocs:Getsc'
                            b'reen:Background_Black.png\x00\x00\x0e'
                            b'\x00*\x00\x14\x00B\x00a\x00c\x00k\x00g\x00r'
                            b'\x00o\x00u\x00n\x00d\x00_\x00B\x00l\x00a'
                            b'\x00c\x00k\x00.\x00p\x00n\x00g\x00\x0f\x00\x1a'
                            b'\x00\x0c\x00M\x00a\x00c\x00i\x00n\x00t\x00o'
                            b'\x00s\x00h\x00 \x00H\x00D\x00\x12\x00WUsers/gwen'
                            b'd/Library/Mobile Documents/com~apple~CloudDocs/G'
                            b'etscreen/Background_Black.png\x00\x00\x13'
                            b'\x00\x01/\x00\x00\x15\x00\x02\x00\x0c\xff\xff'
                            b'\x00\x00',
    'backgroundType': 0,
    'gridOffsetX': 0.0,
    'gridOffsetY': 0.0,
    'gridSpacing': 100.0,
    'iconSize': 48.0,
    'labelOnBottom': True,
    'showIconPreview': True,
    'showItemInfo': False,
    'textSize': 12.0,
    'viewOptionsVersion': 1
}
.                pBBk Bookmark([(0x1, {
  0x1004: ['Users', 'gwend', 'Library', 'Mobile Documents', 'com~apple~CloudDocs', 'Getscreen', 'Background_Black.png']
  0x1005: [702312, 1096918, 1096919, 1101927, 1101963, 4487648, 4476282]
  0x1010: Data(b'\x01\x00\x00\x00\x00\x00\x00\x00\x0f\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00')
  0x1040: datetime.datetime(2019, 10, 13, 12, 26, 6, tzinfo=<mac_alias.utils.UTC object at 0x8015ce490>)
  0x2002: '/'
  0x2005: URL('file:///')
  0x2010: 'Macintosh HD'
  0x2011: '81DED881-7FF2-4FBB-A8F8-DC8A4BC4C10E'
  0x2012: 250685575168
  0x2013: datetime.datetime(2020, 5, 23, 16, 54, 30, 286462, tzinfo=<mac_alias.utils.UTC object at 0x8015ce490>)
  0x2020: Data(b'\x81\x00\x00\x00\x01\x00\x00\x00\xef\x13\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00')
  0x2030: True
  0xc001: 5
  0xc011: 'gwend'
  0xc012: 501
  0xd001: True
  0xd010: 536871424
  'NSURLDocumentIdentifierKey': 306
}),
])
.                vSrn 1
Applications     Iloc (268, 64)
Getscreen.me.app Iloc (88, 64)

ds_store testdata/00.DS_Store  0,08s user 0,01s system 99% cpu 0,089 total
```

Takes around 80ms; not all that bad?
We could also "cache" the relevant key-value pairs in extended attributes to speed things up for subsequent invocations...

**Comment by moochris** [comment]
Yeah that's not too bad. I guess we'd only be reading, and then only once because we'd then store in the new extattrs and (optionally) dot files after that?

**Comment by probonopd** [comment]
No speed gain using the Nuitka compiler, I am surprised:

```
% sudo pkg install nuitka-py37
% python3 -mnuitka --follow-imports /usr/local/bin/ds_store
%  time ./ds_store.bin DS_Store
./ds_store.bin DS_Store  0,09s user 0,02s system 99% cpu 0,109 total
```

**Comment by probonopd** [comment]
> Yeah that's not too bad. I guess we'd only be reading, and then only once because we'd then store in the new extattrs and (optionally) dot files after that? (Right, essentially what you said in the last paragraph now I read it again slightly_smiling_face)

Ideally we would be using extended attributes as the "cache", and write back to proper `.DS-Store` files when the time comes to write a dotfile. The goal being not to need yet another dotfile that some people tend to hate and need to take care of (e.g., in `.gitignore` and so on).

**Comment by probonopd** [comment]
Cound not get cython to work yet:

```
% cython-3.7 ds_store.pyx --embed                   
/usr/local/lib/python3.7/site-packages/Cython/Compiler/Main.py:369: FutureWarning: Cython directive 'language_level' not set, using 2 for now (Py2). This will change in a later release! File: /usr/home/user/go/src/github.com/gwend/dsstore/ds_store.pyx
  tree = Parsing.p_module(s, pxd, full_module_name)

Error compiling Cython file:
------------------------------------------------------------
...
    for path in args.paths:
        if os.path.isdir(path):
            path = os.path.join(path, ".DS_Store")

        if not os.path.exists(path) or not os.path.isfile(path):
            print('ds_store: {} not found'.format(path), file=sys.stderr)
                                                            ^
------------------------------------------------------------

ds_store.pyx:62:61: Expected ')', found '='

```

**Comment by moochris** [comment]
Yes I agree about not writing another dot file format, but a bit hesitant about whether we'll be able to exactly replicate the spatial metadata - a lot of it would be implementation dependent on Finder I expect. E.g. positions of windows/icons.

So at least IMO I think we shouldn't be focussed on interoperability with OS X/macOS, just reuse the same storage mechanism?

**Comment by probonopd** [comment]
> So at least IMO I think we shouldn't be focussed on interoperability with OS X/macOS

In fact I think we should at least try, because:

* It would be beyond cool if we could get it to be interoperable
* Someone already has thought about the format and the parser
* Looking at the above, most keys/values don't seem to be super cryptic, hard, or unexpected

If it turns out to be too much hassle we can always fall back to custom key/value pairs, but why not use `Iloc` to store the icon location as an x, y tuple?

**Comment by moochris** [comment]
Ok, so far we just store and retrieve the window position and size for a folder. If we use your example, presumably this is:

'WindowBounds': '{{200, 458}, {360, 222}}'

So is that an X,Y pair of the top left and bottom right corners? I guess Y has an origin at the bottom of the screen (opposite to our Qt coordinates), otherwise that wouldn't make sense... This is just a guess anyway, is this documented as well do you know?

**Comment by probonopd** [comment]
Since I have access to a Mac I can try. Let's make a list what I should find out.

**Comment by probonopd** [comment]
Hold on...
> kCGWindowBounds is a field in a dictionary returned by the CGWindowListCopyWindowInfo function. This fields lists the dimensions of a window displayed by the Quartz Window System (Mac OS X’s analogue of X11).
>
> kCGWindowBounds measures the origin as the top-left corner of the rectangle relative to the top-left corner of the screen
> NSWindow’s frame property measures the origin as the bottom-left corner of the rectangle relative to the bottom-left corner of the screen

http://www.krizka.net/2010/04/20/converting-between-kcgwindowbounds-and-nswindowframe/

Wouldn't it make more sense for `kCGWindowBounds` to be used here in order to deal consistently with different window sizes?

In any case, I have access to a Mac and can test stuff out - let's make a list of what I should test.

**Comment by moochris** [comment]
Hmm, ok. Yes I think test with your Mac and see what you get for e.g. putting a window in the top left or bottom left corners.

I haven't had a Mac since my iBook G4 died 🙂

**Comment by probonopd** [comment]
Interesting... in OS X 10.9
* The button in the upper-right hand corner of each Finder window to switch to Spatial Mode is gone (how can I enable it "nowadays"?)
* Just creating a Folder in Finder doesn't generate a `.DS_Store` file

**Comment by probonopd** [comment]
Let's brainstorm a bit on the metadata necessary for spatial mode.

Terms
* Spatial Mode: If the file manager (e.g., Filer) uses Spatial mode rather than Brwoser mode, then a new window is opened for each folder
* Spatial information: Metadata that describes windows (for folders) and how they are shown, and the items inside the window (files, folders, etc.)

Objectives
* Spatial metadata (e.g, window sizes, icon positions, window backgrounds etc.) need to be able to travel inside directories in "robust" ways (e.g., when copied over  a network or when put in an archive using all existing tools)
* Even if the file manager is not using Spatial mode overall, it needs to be possible that individual folders are shown like the file manager was set to Spatial mode. For example, many disk images used for application installations use this
* Provide a spec that can be implemented not only by helloSystem and its file manager, but also by third parties

Mac `.DS_Store` files are able to fulfill the objectives, but are not officially documented, although partial third-party implementations for reading and writing them exist. In line with the helloSystem objectives, we don't necessarily want to reinvent the wheel, but reuse existing working solutions where possible. So one question is whether we could use the existing `.DS_Store` file format for our purposes.

##; Pros of using `.DS_Store`

* Users already know what these files are for and how to handle them (e.g., in `.gitignore` files)
* Less littering: If `.DS_Store` files already exist in a filesystem tree, then no additional files would be needed

##; Cons of using `.DS_Store`

* The format is not officially documented, so implementing it is partially guesswork
* The format is more complex than we would like (e.g., uses Bookmarks/Aliases which are also not officially documented)
* Implementation will be more complex than needed because the format assumes underlying logic (e.g., coordinate systems) to be the NEXTStep/AppKit way, which is unlike Qt

##; Pros of using an own format

* It can be much simpler
* It does not need to deal with NEXTStep/AppKit specifics
* We can shape it
* There can be an official open spec

##; Cons of using an own format

* More littering, especially in environments with macOS and helloSystem there will be two files rather than one
* We need to think about and define the format

Let's start to think about and define the format. This is pure brainstorming at this point to see what we might end up with, and how it would compare to `.DS_Store`.

##; Potential `.DirInfo` format

* In general, every key-value combination can be either saved to a file, `.DirInfo`, in each folder (directory), and/or to extended attributes in the filesystem (if the filesystem supports it)
* __We need to think about whether caching in  extended attributes in the filesystem is actually worthwhile the additional complexity; especially given that one would then always need to check where the latest information is and somehow keep them in sync and/or decide which information "wins". Maybe we can learn from Haiku?__
* Wouldn't that mean that we always have to check the `.DirInfo` (if it exists) in addition to the extended attributes? (Yes, but maybe we write the `.DirInfo` only in exceptional cases when it was explicitly requested?)
* In order not to litter the filesystem more than needed, the information from the extended attributes might be saved to `.DirInfo` files only upon defined trigger points. Which trigger points these would be? (The most extreme case would be to save a .DirInfo file only when the user explcitly requests it, e.g., by using a "Save" menu item, or when a tool such as a compression/disk image tool specifically requests a `.DirInfo` file to be written)
* Some users really don't like `.DS_Store` files and hence will probably not like `.DirInfo` files either. We should define a really simple way to opt out of the creation of `.DirInfo` files; e.g., by the convention that `.DirInfo` files will not be created if the toplevel directory of a volume contains a file `.DisableDirInfo`

We should use a simple key-value format that could be saved to `.DirInfo` files and/or extended attributes.
Would the [.ini format](https://en.wikipedia.org/wiki/INI_file) that is understood by many programming languages and frameworks be suitable for the on-disk representation? It seems to be cumbersome to write by hand "thanks" to excaping rules...
Would JSON, or [UCL](https://papers.freebsd.org/2015/bsdcan/allanjude-ucl/) (a relaxed version of JSON), or YAML be better? (Why is macOS not using `.plist` for `.DS_Store`?)

All keys are optional. If a key is not set or the value is empty, then the file manager default and/or default user setting is used, whatever that may be.

```
[Window]

; These key-value pairs get cached in extended attributes on the directory (folder)

ShowPathbar=False ; Only implement if file manager has a pathbar that is not part of the toolbar
ShowSidebar=False
; ShowStatusBar=False ; Initially not implemented for simplification 
ShowTabView=False
ShowToolbar=False
; SidebarWidth=284.0 ; Initially not implemented for simplification
; WindowBounds= ; Not used due to different coordinate system
WindowX= ; pixels from left
WindowY= ; pixels from top
WindowWidth= ; pixels
WindowHeight= ; pixels
SortBy=Name ; Name, Mtime, Size, Type, Owner
SortDescending=False
SortFoldersFirst=True
SortCaseSensitive=True

; BackgroundImage=.background.png ; Not used for simplification, we standardize on .DirBackground.png and .DirBackground-2x.png (HiDPI)
; BackgroundType=0 ; Not used for simplification
GridOffsetX=0.0
GridOffsetY=0.0
GridSpacing=100.0
IconSize=48.0
; LabelOnBottom=True ; Initially not implemented for simplification
ShowIconPreview=True
; ShowItemInfo=False ; Initially not implemented for simplification
; TextSize=12.0 ; Initially not implemented for simplification
; ViewOptionsVersion=1 ; Initially not implemented for simplification
```

Note that the file manager needs to check whether `WindowX` and `WindowY` lie within the visible space of the displays currently connected to the machine (plus some margin); and use some other values that are within the visible space (plus some margin) otherwise.

It is less clear how to best store the information regarding the individual items. The following may be problematic due to spaces and other special characters in file names which may prove problematic when being used as keys.

```
[Items]

; Should these key-value pairs get cached in extended attributes on the individual files or also in the directory?
; Will we run into issues with special characters and blanks in the keys?
; Should we use another format than the .ini format?

My Application.app=100,100
Other File.txt=200,100
```

In JSON, this mgiht look like:

```
{
    "Window": {
        "ShowPathbar": false,
        "ShowSidebar": false,
        "ShowTabView": false,
        "ShowToolbar": false,
        "WindowX": 100,
        "WindowY": 200,
        "WindowWidth": 300,
        "WindowHeight": 400,
        "SortBy": "Name",
        "SortDescending": false,
        "SortFoldersFirst": true,
        "SortCaseSensitive": true,
        "IconSize": 48,
        "ShowIconPreview": true
    },
    "Items": {
        "My Application.app": {
            "IconX": 100,
            "IconY": 100
        },
        "Other File.txt": {
            "IconX": 200,
            "IconY": 100
        }
    }
}
```

In UCL, the following simplication might be allowable:

```
  Window {
    ShowPathbar = false;
    ShowSidebar = false;
    ShowTabView = false;
    ShowToolbar = false;
    WindowX = 100;
    WindowY = 200;
    WindowWidth = 300;
    WindowHeight = 400;
    SortBy = Name;
    SortDescending = false;
    SortFoldersFirst = true;
    SortCaseSensitive = true;
    IconSize" = 48;
    ShowIconPreview = true;
};

Items {
    "My Application.app" = {
        IconX = 100;
        IconY = 100;
    };
    "Other File.txt" = {
        IconX = 200;
        IconY = 100;
    }
}
```

Since JSON is readily available everywhere (and is a valid form of UCL), is it the sweet spot?
These files will probably not often be written by hand.

In YAML this would be:

```
Window:
  ShowPathbar: false
  ShowSidebar: false
  ShowTabView: false
  ShowToolbar: false
  WindowX: 100
  WindowY: 200
  WindowWidth: 300
  WindowHeight: 400
  SortBy: Name
  SortDescending: false
  SortFoldersFirst: true
  SortCaseSensitive: true
  IconSize: 48
  ShowIconPreview: true
Items:
  My Application.app:
    IconX: 100
    IconY: 100
  Other File.txt:
    IconX: 200
    IconY: 100
```

Neat for editing by hand, but YAML is not natively supported in Qt and would require a third-party library such as https://github.com/jbeder/yaml-cpp which incrases the dependency tree and complexity.

What are your thoughts on these topics?

**Comment by moochris** [comment]
> * Spatial metadata (e.g, window sizes, icon positions, window backgrounds etc.) need to be able to travel inside directories in "robust" ways (e.g., when copied over  a network or when put in an archive using all existing tools)
> * Even if the file manager is not using Spatial mode overall, it needs to be possible that individual folders are shown like the file manager was set to Spatial mode. For example, many disk images used for application installations use this

Yes, there is metadata that isn't necessarily spatial - attributes of files, e.g. tagged colours should show in both modes.

> * Implementation will be more complex than needed because the format assumes underlying logic (e.g., coordinate systems) to be the NEXTStep/AppKit way, which is unlike Qt

For me this is the biggest con of using .DS_Store - we are constrained where we could innovate :)

> ### Pros of using an own format
> * It can be much simpler

Not only simpler, but human readable.

> * It does not need to deal with NEXTStep/AppKit specifics
> * We can shape it
> * There can be an official open spec

This is a very good aim in my opinion.

> * **We need to think about whether caching in  extended attributes in the filesystem is actually worthwhile the additional complexity; especially given that one would then always need to check where the latest information is and somehow keep them in sync and/or decide which information "wins". Maybe we can learn from Haiku?**

I'm pretty sure Haiku _only_ uses BeFS attributes to store this information. Therefore it avoids some things we need to consider if writing both dot files and attributes. In Haiku the attributes are written to both files and folders.

e.g. we might want to give files particular attributes like colours to tag them. If we are storing the file's metadata in it's containing folder's dot file, then when we move the file from one folder to another, we need to 'move' the metadata from one dot file to another. 

> * Wouldn't that mean that we always have to check the `.DirInfo` (if it exists) in addition to the extended attributes? (Yes, but maybe we write the `.DirInfo` only in exceptional cases when it was explicitly requested?)

I think you only read the .DirInfo if the extended attributes are not present - this would only be in the cases where the files/folders have been on a file system or in an archive where they lost the extended attributes.

> * In order not to litter the filesystem more than needed, the information from the extended attributes might be saved to `.DirInfo` files only upon defined trigger points. Which trigger points these would be? (The most extreme case would be to save a .DirInfo file only when the user explcitly requests it, e.g., by using a "Save" menu item, or when a tool such as a compression/disk image tool specifically requests a `.DirInfo` file to be written)

I was thinking that you'd write the file when you close the window. You would have a function that iterated over all the extended attributes for the folder and all files/folders within and constructed the JSON (for example) from that. It won't slow the user down because the window will have disappeared already (as far as they're concerned).

> * Some users really don't like `.DS_Store` files and hence will probably not like `.DirInfo` files either. We should define a really simple way to opt out of the creation of `.DirInfo` files; e.g., by the convention that `.DirInfo` files will not be created if the toplevel directory of a volume contains a file `.DisableDirInfo`

Yes, or an option in the preferences to turn off the .DirInfo files?

> In UCL, the following simplication might be allowable:
> 
> ```
>   Window {
>     ShowPathbar = false;
>     ShowSidebar = false;
>     ShowTabView = false;
>     ShowToolbar = false;
>     WindowX = 100;
>     WindowY = 200;
>     WindowWidth = 300;
>     WindowHeight = 400;
>     SortBy = Name;
>     SortDescending = false;
>     SortFoldersFirst = true;
>     SortCaseSensitive = true;
>     IconSize" = 48;
>     ShowIconPreview = true;
> };
> 
> Items {
>     "My Application.app" = {
>         IconX = 100;
>         IconY = 100;
>     };
>     "Other File.txt" = {
>         IconX = 200;
>         IconY = 100;
>     }
> }
> ```
> 
> Since JSON is readily available everywhere (and is a valid form of UCL), is it the sweet spot?
> These files will probably not often be written by hand.

I really like the UCL format. The extended attributes we are writing/reading already are:

```
static const QString WINDOW_ORIGIN_X            = "window.originx";
static const QString WINDOW_ORIGIN_Y            = "window.originy";
static const QString WINDOW_HEIGHT              = "window.height";
static const QString WINDOW_WIDTH               = "window.width";
```

So the UCL representation is already along those lines - you'd flatten them like that for the extended attributes.

**Comment by moochris** [comment]
Now I think about it, saving on window close won't be enough - a user could move things in a window that's open, and then from the parent folder drag it to a network drive...

So at least on window close, but also on some other triggers?

**Comment by probonopd** [comment]
One more thought, since those files will be mainly written automatically (rather than by hand), would using a __binary__ format optimized for speed be an advantage? I suspect there is a reason why _binary_ plists exist.

**Comment by moochris** [comment]
Yes, I would expect it would be faster. If it's fast enough, we could just write to the dot file at the same time as extended attributes

**Comment by probonopd** [comment]
> binary format optimized for speed

Looks like [MessagePack](https://msgpack.org/) is such a protocol.

> MessagePack is an efficient binary serialization format. It lets you exchange data among multiple languages like JSON. But it's faster and smaller. Small integers are encoded into a single byte, and typical short strings require only one extra byte in addition to the strings themselves.

https://github.com/romixlab/qmsgpack

> There is packers and unpackers for QColor, QTime, QDate, QDateTime, QPoint, QSize, QRect. Also you can create your own packer/unpacker methods for Qt or your own types.

Not sure whether using something like this would be premature optimization for our use case. Maybe we shoud start with something like JSON and check what the performance is.

**Comment by moochris** [comment]
I think in parallel we should start a new filer implementation from scratch, with only Qt dependencies (perhaps leveraging FreeBSD or ZFS features along the way).

I think adapting existing projects will get us so far, but never to exactly what we want.

The spatial/metadata spec is of course useful for doing this.

**Comment by probonopd** [comment]
> I think in parallel we should start a new filer implementation from scratch, with only Qt dependencies (perhaps leveraging FreeBSD or ZFS features along the way).

All the power to you, but I assume it'd be a massive undertaking.

Maybe we could leverage one of the existing ones that have been mentioned, like DFileManager or https://github.com/cyberos/cyber-fm which in turn is based on https://invent.kde.org/maui/index-fm. cyber-fm and index-fm are written in Qml though, I am not sure yet whether this is a great match for a file manager but I guess it is worth some experimentation.

**Comment by moochris** [comment]
Yes, you might be right 🙂

Ok, certainly some worth investigating there.

**Comment by moochris** [comment]
Yes, you might be right 🙂

Ok, certainly some worth investigating there.

**Comment by probonopd** [comment]
Let's take on manageable changes. Like https://github.com/helloSystem/Filer/issues/9 - things like that woud already make a _huge_ differece and can be in a small fraction of the time compared to a whole new file manager.

**Comment by moochris** [comment]
> Not sure whether using something like this would be premature optimization for our use case. Maybe we shoud start with something like JSON and check what the performance is.

Interestingly, I was just looking at an example for JSON support in Qt here:
https://doc.qt.io/qt-5/qtcore-serialization-savegame-example.html

"With [QJsonDocument](https://doc.qt.io/qt-5/qjsondocument.html), you also have the ability to serialize a document in a [CBOR](https://tools.ietf.org/html/rfc7049) format, which is great if you don't want the save file to be readable, or if you need to keep the file size down"

I wonder if saving to CBOR is also quicker?

**Comment by probonopd** [comment]
Interesting, I had never heard of it. Maybe we should try JSON and then see whether additional optimization is needed.

**Comment by probonopd** [comment]
> Yes, or an option in the preferences to turn off the .DirInfo files?

There needs to be a way to disable the creation of `.DirInfo` files for a certain volume (mount point). E.g., think of a network share where the owner wants no `.DirInfo` files, regardless of what various users may have set up in Filer.

**Comment by moochris** [comment]
> There needs to be a way to disable the creation of `.DirInfo` files for a certain volume (mount point). E.g., think of a network share where the owner wants no `.DirInfo` files, regardless of what various users may have set up in Filer.

OK - do you think we need to walk up each level in the path and check for a `.DisableDirInfo` file? Just thinking they might not necessarily have them at the mount point.

**Comment by probonopd** [comment]
I'd say who owns the mountpoint decides ;-) so let's not make it overly complex and slow.

**Comment by grahamperrin** [comment]
[On the origins of DS_store](https://www.arno.org/on-the-origins-of-ds-store)

> … I was the technical lead for the Mac OS X Finder …

**Comment by probonopd** [comment]
When using Spatial mode, I urgently feel the need to have a "Go to... Ctrl+G" dialog similar to

![image](https://user-images.githubusercontent.com/2480569/111075360-fbb22500-84e7-11eb-9bd9-516ec777570a.png)

**Comment by grahamperrin** [comment]
The wish to enter a location is not limited to spatial mode. 

For a non-swapped keyboard layout, <kbd>Control</kbd>-<kbd>L</kbd> will be ideal.

**Comment by probonopd** [comment]
> The wish to enter a location is not limited to spatial mode.

Indeed. 

> Control-L

L? Why L? Command+Shift+G is it... in line with https://github.com/helloSystem/hello/issues/58

https://osxdaily.com/2011/08/31/go-to-folder-useful-mac-os-x-keyboard-shortcut/

**Comment by probonopd** [comment]
@moochris do you think you can hook this in?

Go -> Go To Folder... Ctrl+Shift+G

```
    QWidget *centralWidget = new QWidget;

    QVBoxLayout *layout = new QVBoxLayout;

    label = new QLabel;
    label->setText("Go to the folder:");
    layout->addWidget(label);

    lineEdit = new QLineEdit;
    lineEdit->setMinimumWidth(480);
    layout->addWidget(lineEdit);

    buttonBox = new QDialogButtonBox;
    buttonBox->addButton(tr("Go"), QDialogButtonBox::AcceptRole);
    buttonBox->addButton(tr("Cancel"),QDialogButtonBox::RejectRole);

    // TODO: Implement functionality

    layout->addWidget(buttonBox);

    centralWidget->setLayout(layout);
    setCentralWidget(centralWidget);

    setWindowTitle(tr("Go To Folder"));
    lineEdit->setFocus();

    completer = new QCompleter(this);
    completer->setCompletionMode(QCompleter::InlineCompletion);
    QFileSystemModel *fsModel = new QFileSystemModel(completer);
    fsModel->setFilter(QDir::Dirs|QDir::Drives|QDir::NoDotAndDotDot|QDir::AllDirs); // Only directories, no files
    completer->setModel(fsModel);
    fsModel->setRootPath(QString());
    lineEdit->setCompleter(completer);
```

**Comment by moochris** [comment]
Nice - yep, leave it with me and I'll hook it up.

I'm moving home this weekend so can't guarantee I'll get to it quickly.

**Comment by moochris** [comment]
@probonopd  PR for it here: https://github.com/helloSystem/Filer/pull/56

**Comment by KDGNOR** [comment]
some text about adding colors tags...

here : https://github.com/helloSystem/Filer/issues/32

**Comment by probonopd** [comment]
Spatial file manager including storing icon position is being implemented in https://github.com/probonopd/Filer/.
